web-components/src/components/table-advanced/TableAdvanced.ts

```diff
@@ -0,0 +1,994 @@
+/**
+ * Copyright (c) Cisco Systems, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ */
+
+import reset from "@/wc_scss/reset.scss";
+import "@/components/icon/Icon";
+import "@/components/menu-overlay/MenuOverlay";
+import "@/components/button/Button";
+import { html, internalProperty, LitElement, property, PropertyValues } from "lit-element";
+import styles from "./scss/module.scss";
+import { ifDefined } from "lit-html/directives/if-defined";
+import { classMap } from "lit-html/directives/class-map";
+import { templateContent } from "lit-html/directives/template-content";
+import { nothing, TemplateResult } from "lit-html";
+import Papa from "papaparse";
+import { customElementWithCheck } from "@/mixins/CustomElementCheck";
+import { Filter } from "./src/filter";
+import { debounce, Evt, evt, TemplateCallback, templateCallback, TemplateInfo } from "./src/decorators";
+
+const IMG = document.createElement("img");
+IMG.src = "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7";
+
+export namespace TableAdvanced {
+  /**
+   * @element md-table-advanced
+   * @fires md-table-advanced-change
+   */
+  @customElementWithCheck("md-table-advanced")
+  export class ELEMENT extends LitElement {
+    @property({ attribute: false }) config!: Config;
+    @property({ attribute: false }) data!: Data;
+
+    @evt() "md-table-advanced-change"!: Evt<ChangeEvent>;
+
+    @internalProperty() private error = "";
+
+    @internalProperty() private COLS: Col[] = [];
+    @internalProperty() private ROWS: Cell[][] = [];
+    private updCols = () => this.requestUpdate("COLS");
+    private updRows = () => this.requestUpdate("ROWS");
+
+    @internalProperty() private dragRow = -1;
+    @internalProperty() private dropRow = -1;
+    @internalProperty() private drops: [number, number][] = [];
+    private dragRowElem: HTMLElement | null = null;
+
+    @internalProperty() private dragCol = -1;
+    @internalProperty() private dropCol = -1;
+
+    private isResizing = false;
+    private isSelectable = false;
+    private selected: Record<number, boolean> = {};
+    private expandedRowIdx: Record<number, boolean> = {};
+
+    private dragover = (e: Event) => e.preventDefault();
```

Comment:
@futurethang internalProperty will change template(render) automatically
in case of "selected" or "expandedRowIdx" I know that I should "requestUpdate()" if needed

Reply:
@futurethang they are consistent/organized by "context/feature" not by declaration form